File: src/main/scala/com/bjsxt/sparkscalacode/SparkGroup1.scala
package com.bjsxt.sparkscalacode

import org.apache.spark.sql.SparkSession
import org.apache.spark.{SparkConf, SparkContext}

import scala.util.control.Breaks

object SparkGroup1 {
  def main(args: Array[String]): Unit = {
    val sc = new SparkContext(new SparkConf().setMaster("local").setAppName("test"))
    /*val test: SparkSession.Builder = SparkSession.builder().master("local").appName("test")
    val session: SparkSession = SparkSession.builder().master("local").appName("test").getOrCreate()
    val context: SparkContext = session.sparkContext*/
    //val sc = SparkSession.builder().master("local").appName("test").getOrCreate().sparkContext

    val lines = sc.textFile("./data/scores.txt")

    lines.map(s=>(s.split("\t")(0),s.split("\t")(1).toInt)).groupByKey().foreach(s=>{
      val key = s._1
      val iter = s._2.iterator
      val ints = new Array[Int](3)

      val loop = new Breaks
      while (iter.hasNext){
        val next = iter.next()
        loop.breakable(
          for (i <- 0 until ints.length){
            if (next>ints(i)){
              for (j <- (ints.length-1).until(i,-1)){
                ints(j) = ints(j-1)
              }
              ints(i) = next
              loop.break()
            }
          }
        )
      }
      println(s"class is $key")
      ints.foreach(println)


      /*val loop = new Breaks
      while (iter.hasNext){
        val next = iter.next()
        loop.breakable()
        for (i <- 0 until ints.length){
          if (next>ints(i)){
            for (i <- ints.length-1 until i){
              ints(i) = ints(i-1)
            }
            ints(i) = next
          }
        }
      }*/


      /*//原生集合排序
      val iterable: Iterable[Int] = s._2
      val list = iterable.toList
      val listSort = list.sortWith(_>_).take(3)
      println(s"class is $key")
      listSort.foreach(println)*/
    })
  }
}

File: src/main/java/com/bjsxt/sparkjavacode/SparkGroup1.java
package com.bjsxt.sparkjavacode;

import com.sun.org.apache.bcel.internal.generic.SWAP;
import org.apache.commons.collections.IteratorUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import scala.Array;
import scala.Int;
import scala.Tuple2;

import java.util.*;

public class SparkGroup1 {
    public static void main(String[] args) {
        JavaSparkContext sc = new JavaSparkContext(new SparkConf().setMaster("local").setAppName("test"));

        JavaRDD<String> lines = sc.textFile("./data/scores.txt");

        lines.mapToPair(s->new Tuple2<>(s.split("\t")[0],Integer.valueOf(s.split("\t")[1]))).
                groupByKey().foreach(s->{
            String key = s._1;
            Iterator<Integer> iter = s._2.iterator();
            int[] ints = new int[3];
            while (iter.hasNext()){
                Integer next = iter.next();
                for (int i = 0; i < ints.length; i++) {
                    if (next>ints[i]){
                        for (int j = ints.length-1; j > i; j--) {
                            ints[j] = ints[j-1];
                        }
                        ints[i] = next;
                        break;
                    }
                }
            }
            for (Integer i :
                    ints) {
                System.out.println("key = "+key+" value = "+i);
            }

            /*//原生集合排序

//            Iterable<Integer> iter = s._2;
//            ArrayList<Integer> list = new ArrayList<>();
//            for (Integer i:iter){
//                list.add(i);
//            }
            Iterator<Integer> iter = s._2.iterator();
            List<Integer> list = IteratorUtils.toList(iter);
            Collections.sort(list, new Comparator<Integer>() {
                @Override
                public int compare(Integer o1, Integer o2) {
                    return o2-o1;
                }
            });
            for (Integer i : list){
                System.out.println("key = "+key+" value = "+i);
            }*/
        });

        System.out.println("+++++++++++++++++++++++");
        lines.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<>(s.split("\t")[0],Integer.valueOf(s.split("\t")[1]));
            }
        }).groupByKey().foreach(s->{

            //原生集合
            List list = IteratorUtils.toList(s._2.iterator());
            Collections.sort(list, new Comparator<Integer>() {
                @Override
                public int compare(Integer o1, Integer o2) {
                    return o2-o1;
                }
            });
            int i = 0;
            System.out.println("class = "+s._1);
            for (Iterator iter2 = list.iterator(); iter2.hasNext()&&i<3 ; i++) {
                System.out.println(iter2.next());
            }
        });
    }
}
